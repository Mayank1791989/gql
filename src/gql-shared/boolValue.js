/* @flow */
export default function boolValue(
  value: ?boolean,
  defaultValue: boolean,
): boolean {
  return typeof value === 'boolean' ? value : defaultValue;
}
