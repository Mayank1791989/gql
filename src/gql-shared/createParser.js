/* @flow */
import { type IParser } from './types';

export default function createParser(
  params: [Class<IParser>, Object],
): IParser {
  const [Parser, options] = params;
  return new Parser(options);
}
