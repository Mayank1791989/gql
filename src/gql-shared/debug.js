/* @flow */
import noop from './noop';

type Debug = { enable: () => void } & typeof console;

const debug: Debug = new Proxy(console, {
  _isEnabled: false,
  get(target, key) {
    if (key === 'enable') {
      this._isEnabled = true;
      return noop;
    }
    return this._isEnabled ? target[key] : noop;
  },
});

export default debug;
