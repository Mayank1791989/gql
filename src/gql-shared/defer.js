/* @flow */
export default function defer<T>(): {
  promise: Promise<T>,
  resolve: T => void,
  reject: () => void,
} {
  const deferred = {};
  deferred.promise = new Promise<T>((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });
  return deferred;
}
