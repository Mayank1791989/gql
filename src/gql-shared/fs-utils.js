/* @flow */
import mkdirp from 'mkdirp';
import fs from 'fs';
import path from 'path';

export function mkdirpSync(dir: string) {
  mkdirp.sync(dir, { fs });
}

export function writeFileSync(file: string, content: string) {
  const dir = path.dirname(file);
  // make sure dir exists
  mkdirpSync(dir);
  fs.writeFileSync(file, content);
}
