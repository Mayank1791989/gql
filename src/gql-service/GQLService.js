/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log from 'gql-shared/log';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';
import getResolverLocation from 'gql-schema-service/commands/getResolverLocation';
import getSchemaLocation from 'gql-schema-service/commands/getSchemaLocation';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

type GetResolverOrSchemaParams = $ReadOnly<{|
  fieldType: string,
  fieldName: string,
|}>;

export type { Options as GQLServiceOptions };

export function isGql(sourcePath: string) {
  return !(sourcePath.endsWith('ts') || sourcePath.endsWith('ts'));
}

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService: this._schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  onLog(listener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      logger.time('autocomplete response');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getHintsAtPosition');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      logger.timeEnd('getHintsAtPosition');
      logger.timeEnd('autocomplete response');
      return results;
    }, []);
  }

  getLenses(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getLenses request');
      logger.time('getLenses response');
      if (!this._isRunning) {
        return null;
      }

      let defLenses = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getLenses');

        if (fileConfig && fileConfig.type === 'schema') {
          defLenses = await this._schemaService.getLenses({
            filePath: params.sourcePath,
          });
        }
      } else {
        logger.time('getLenses');
        defLenses = await this._schemaService.getLenses({
          filePath: params.sourcePath,
        });
      }

      logger.timeEnd('getLenses');
      logger.timeEnd('getLenses response');
      return defLenses;
    }, null);
  }

  getResolverLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getResolverLocation request');
      logger.time('getResolverLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getResolverLocation');

      const schema = this._schemaService.getSchema();
      const resolverLocation = await getResolverLocation(
        params.fieldType,
        params.fieldName,
        schema,
        this._config.getResolverConfig().files,
        this._config.getResolverConfig().baseDir,
      );

      logger.timeEnd('getResolverLocation');
      logger.timeEnd('getResolverLocation response');
      return resolverLocation;
    }, null);
  }

  getSchemaLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      logger.debug('getSchemaLocation request');
      logger.time('getSchemaLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getSchemaLocation');

      const schema = this._schemaService.getSchema();
      const schemaLocation = getSchemaLocation(
        params.fieldType,
        params.fieldName,
        schema,
      );

      logger.timeEnd('getSchemaLocation');
      logger.timeEnd('getSchemaLocation response');
      return schemaLocation;
    }, null);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getDef request');
      logger.time('getDef response');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getDef');
        if (fileConfig && fileConfig.type === 'schema') {
          defLocation = await this._schemaService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
            resolversGlob: this._config.getResolverConfig().files,
            resolversBaseDir: this._config.getResolverConfig().baseDir,
          });
        }

        if (this._queryService && fileConfig && fileConfig.type === 'query') {
          defLocation = await this._queryService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
          });
        }
      } else {
        logger.time('getDef');
        defLocation = await this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getDef');
      logger.timeEnd('getDef response');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      logger.time('findRefs response');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      // if (this._queryService && fileConfig && fileConfig.type === 'query') {
      //   refLocations = this._queryService.findRefsOfTokenAtPosition({
      //     fileContent: params.sourceText,
      //     filePath: params.sourcePath,
      //     fileOptions: fileConfig.opts,
      //     position: params.position,
      //   });
      // }
      logger.timeEnd('findRefs');
      logger.timeEnd('findRefs response');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      logger.time('getInfo response');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getInfoAtToken');
      logger.timeEnd('getInfo response');
      return info;
    }, null);
  }
}
