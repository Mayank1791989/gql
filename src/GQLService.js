/* @flow */
import * as schema from './schema';
import * as query from './query';

import GQLConfig from './config/GQLConfig';
import debug from './shared/debug';

import type {
  GQLHint,
  GQLInfo,
  DefLocation,
  Position,
} from './shared/types';

import type { GQLError } from './shared/GQLError';

type Options = $Exact<{
  cwd?: string,
  onChange?: () => void,
  onInit?: () => void,
  debug?: true,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: Position,
}>;

export type { Options as GQLServiceOptions };

export class GQLService {
  _isInitialized: boolean = false;
  _schemaBuilder: schema.SchemaBuilder;
  _queryManager: ?query.QueryManager;

  _config: GQLConfig;

  constructor(options: ?Options) {
    const { onChange, onInit, debug: enableDebug, ...configOptions } = options || {};
    if (enableDebug) { (debug: any).enable(); }
    this._config = new GQLConfig(configOptions);
    this._schemaBuilder = new schema.SchemaBuilder({
      config: this._config,
      watch: true,
      onChange,
      onInit: () => {
        if (this._config.getQuery()) {
          this._queryManager = new query.QueryManager({
            config: this._config,
            getSchema: () => this._schemaBuilder.getSchema(),
            onInit: () => {
              this._isInitialized = true;
              if (onInit) { onInit(); }
            },
            onChange,
          });
        } else {
          this._isInitialized = true;
          if (onInit) { onInit(); }
        }
      },
    });
  }

  // list all file extensions defined in gqlconfig
  getFileExtensions(): Array<string> {
    return this._config.getFileExtensions();
  }

  status(): Array<GQLError> {
    if (!this._isInitialized) { return []; }
    const schemaErrors = this._schemaBuilder.getSchemaErrors();
    const queryErrors = this._queryManager ? this._queryManager.getErrors() : [];
    return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    debug.log('\n[autocomplete]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return []; }

    // codemirror instance
    let results = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('autocomplete');
    if (match && match.type === 'schema') {
      results = schema.commands.getHintsAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      results = query.commands.getHintsAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('autocomplete');
    debug.timeEnd('time');
    return results;
  }

  getDef(params: CommandParams): ?DefLocation {
    debug.log('\n[getDef]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return undefined; }

    let defLocation = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getDef');
    if (match && match.type === 'schema') {
      defLocation = schema.commands.getDefinitionAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      defLocation = query.commands.getDefinitionAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts.parser,
      );
    }
    debug.timeEnd('getDef');
    debug.timeEnd('time');
    return defLocation;
  }

  findRefs(params: CommandParams): Array<DefLocation> {
    debug.log('\n[findRefs]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return []; }

    let refLocations = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('findRefs');
    if (match && match.type === 'schema') {
      refLocations = schema.commands.findRefsOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }
    debug.timeEnd('findRefs');

    debug.timeEnd('time');
    // @TODO query not implemented
    return refLocations;
  }

  getInfo(params: CommandParams): ?GQLInfo {
    debug.log('\n[getInfo]'); debug.time('time');
    const { sourcePath, sourceText, position } = params;
    if (!this._isInitialized) { return undefined; }

    let info = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getInfoAtToken');
    if (match && match.type === 'schema') {
      info = schema.commands.getInfoOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      info = query.commands.getInfoOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('getInfoAtToken');

    debug.timeEnd('time');
    return info;
  }
}
