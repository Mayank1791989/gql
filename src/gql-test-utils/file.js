/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from 'gql-shared/fs-utils';
import { execSync } from 'child_process';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  execSync(`rm -rf ${TEST_TEMP_DIR}`);
}
