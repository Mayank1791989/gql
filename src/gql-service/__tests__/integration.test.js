/* @flow */
import GQLService from '../GQLService';
import { createTempFiles } from 'gql-test-utils/file';
import { writeFileSync } from 'gql-shared/fs-utils';
import path from 'path';
import invariant from 'invariant';

it('revalidate query files on schema change', async () => {
  const dir = createTempFiles({
    '.gqlconfig': `
      {
        schema: {
          files: 'schema/*.gql',
        },
        query: {
          files: [
            {
              match: 'query/*.graphql',
              presets: ['default'],
            }
          ]
        }
      }
    `,
    'schema/query.gql': `
      type Query {
        viewer: Viewer!
      }
    `,
    'schema/viewer.gql': `
      type Viewer {
        xid: String!
      }
    `,
    'query/test.graphql': `
      query Test {
        viewer {
          id
        }
      }
    `,
  });

  const gql = new GQLService({
    configDir: dir,
    watch: true,
  });
  gql.onError(err => {
    throw err;
  });
  await gql.start();

  // unknown id field error
  expect(gql.status()).toMatchSnapshot();

  // fix error
  writeFileSync(
    path.join(dir, 'schema/viewer.gql'),
    `
      type Viewer {
        id: String!
      }
    `,
  );

  // error should be fixed now
  const errors = await new Promise(resolve => {
    invariant(gql._queryService, 'expected queryService to be defined here');
    gql._queryService.onChange(() => {
      resolve(gql.status());
    });
  });

  expect(errors).toEqual([]);

  // stop watcher and service
  await gql.stop();
});
