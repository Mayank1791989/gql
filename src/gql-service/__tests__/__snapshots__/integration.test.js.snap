// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`revalidate query files on schema change 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 11,
        "line": 4,
        "path": "/tmp/gql-test/d754fff4d57a113e4ea7117fb9888775e4c2821c/query/test.graphql",
      },
    ],
    "message": "Cannot query field \\"id\\" on type \\"Viewer\\". Did you mean \\"xid\\"? (FieldsOnCorrectType)",
    "severity": "error",
  },
]
`;
